docopt_value.h

```diff
@@ -292,20 +308,68 @@ namespace docopt {
 		return variant_.longValue;
 	}
 
+	inline
+	long value::asLongOr(const long value) const noexcept
+	{
+		// Attempt to convert a string to a long
+		if (isString()) {
+			const std::string& str = variant_.strValue;
+			char * str_end;
+			const long ret = std::strtol(str.c_str(), &str_end, 10);
+			// Case 1: No Conversion was possible, return given default value
+			if(str_end == str.c_str() && ret == 0)
+			{
+				return value;
+			}
+			// Case 2: Parsed value is out of range, return given default value
+			if (errno == ERANGE) {
+				return value;
+			}
+			// Case 3: Everything is fine, return parsed value
+			return ret;
+		}
+		if (!isLong())
+		{
+			return value;
+		}
+		return variant_.longValue;
+	}
+
```

Comment:
Could we maybe put this in a private function to be shared with both? That maybe returns std::optional<long>, and one would throw and the other would not?

Reply:
Sorry for my late answer. As far as I know `std::optional` was added to the STL with `C++17` so using it here would mean to either raise the required `C++` version from `C++11` to `C++17` or having multiple implementations depending on the C++ version. I think neither is desirable.

What I could see working is either use `boost::optional` as `boost` is already a dependency because of `boost::regex` or using [SFINAE](https://en.cppreference.com/w/cpp/language/sfinae) to abstract the optional container away like I did in this "simple" example I wrote here https://godbolt.org/z/YPejv9Y71 
(gist mirror: https://gist.github.com/Dadie/d06ea6aca53d37142577b4d1ce305246 )

The later would also enable the usage of other container classes aside from `boost::optional` like `std::optional` but probably introduce unnecessary complexity and might lead to template code explosion. While the former might break some projects depending on docopt as they may not have boost::optional available on their platform.

While I'm not the maintainer of this library (so it's luckily not me who should decide on it) I'm sort of reluctant in neither solution but am open for opinions.

Half-OT:
On another note I'm quite taken by the idea of porting this library to `C++20`. I think this type of library would greatly benefit from newer language features like `consteval` and newer STL classes like `std::string_view` and/or `std::variant`. But (sadly a big but) adoption rate of modern C++ is slow and I would guess it will take at least another 5-8 years till we can assume `C++20` being the baseline. So switching now to a newer standard would do nothing aside from breaking a lot of projects and making the library literally unusable for a lot of people.